### ISSUE TITLE
Cannot translate Contains for simple string comparison (not a collection)

### ISSUE THREAD

**Issue by MichalLechowski** [dotnet/efcore#18741]
After upgrading to 3.0 method using Where with string.Contains(string) inside throws translation error.

### Steps to reproduce
```

 Context.Entity.Where(x => x.Value.ToLower().Contains(searchFilter))
                       .AsQueryable();
```

x is an object with properties int Id and string Value.
searchFilter is a simple string.

Exception similiar to the issue when Contains is called on Collections (saw this issue already reported):

> could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to either AsEnumerable(), AsAsyncEnumerable(), ToList(), or  ToListAsync()

Of course it was working fine before upgrading to 3.0. So is there a way to write it as translateable to SQL or do I have to call the query without this filter and then filter it locally?

### Further technical details

EF Core version: 3.0
Database provider: Microsoft.EntityFrameworkCore.SqlServer (MSSQL 2016)
Target framework: .NET Core 3.0
Operating system:
IDE: Visual Studio 2019 16.3.7

**Comment by ajcvickers** [comment]
@maumar to investigate.

**Comment by smitpatel** [comment]
Possibly caused by #18486

**Comment by MichalLechowski** [comment]
@smitpatel I've checked this out and if your suggestion is equality comparer issues then it's not it.

**Comment by roji** [comment]
@MichalLechowski as we made changes in this area (#18486), can you please try with 3.1.0-preview2 and report the results?

Regardless, I couldn't repro the issue on 3.0 with the following code. If your issue persists with 3.1.0-preview2, could you please submit a full repro such as the following?

```c#
class Program
{
    static void Main(string[] args)
    {
        using var ctx = new BlogContext();
        ctx.Database.EnsureDeleted();
        ctx.Database.EnsureCreated();

        var searchFilter = "f";
        var result = ctx.SomeEntities
            .Where(x => x.Value.ToLower().Contains(searchFilter))
            .ToList();
    }
}

public class BlogContext : DbContext
{
    public DbSet<SomeEntity> SomeEntities { get; set; }

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        => optionsBuilder.UseSqlServer(...);
}

public class SomeEntity
{
    public int Id { get; set; }
    public string Value { get; set; }
}
```

This produces the following SQL (again on 3.0):

```sql
SELECT [s].[Id], [s].[Value]
FROM [SomeEntities] AS [s]
WHERE ((@__searchFilter_0 = N'') AND @__searchFilter_0 IS NOT NULL) OR (CHARINDEX(@__searchFilter_0, LOWER([s].[Value])) > 0)
```

**Comment by AbrahamLopez10** [comment]
Hi @roji , I encountered a issue very similar to this one. I'm running .NET Core 3.0 and EF Core 3.0.

Basically, code like the one below fails saying the LINQ expression cannot be converted to SQL and that I need to use `.AsEnumerable()`:

```cs
private ICollection<App> GetAppsDictionary(ICollection<string> appNames)
{
    return _dbContext.Set<App>.Where(x => appNames.Contains(x.Name))
        .ToDictionary(x => x.Name, x => x);
}
```

In the code above, if I change `appNames` from `ICollection<string>` to just `string[]` then the error is gone and the LINQ expression is translated to SQL.

I'd have imagined that EF Core would be smart enough to translate `ICollection` just like any other enumerable? This works correctly in EF 6 but for some reason doesn't work in EF Core?

**Comment by roji** [comment]
@AbrahamLopez10 can you please try 3.1.0-preview2 as I suggested above?

**Comment by AbrahamLopez10** [comment]
@roji Sorry, thought that recommendation was recommended because the specific issue with `.Value`. I'll try the 3.1.0-preview2 release and report back.

**Comment by roji** [comment]
@AbrahamLopez10 I understand. We also fixed #17342 which seems related to your issue. However, I just checked and that fix wasn't merged in time for 3.1.0-preview2, could you please test with the latest preview3 [from our daily builds](https://github.com/aspnet/AspNetCore/blob/master/docs/DailyBuilds.md)?

**Comment by ajcvickers** [comment]
**EF Team Triage:** Closing this issue as the requested additional details have not been provided and we have been unable to reproduce it.

*BTW this is a canned response and may have info or details that do not directly apply to this particular issue. While we'd like to spend the time to uniquely address every incoming issue, we get a lot traffic on the EF projects and that is not practical. To ensure we maximize the time we have to work on fixing bugs, implementing new features, etc. we use canned responses for common triage decisions.*
